Modules/selectmodule.c

```diff
@@ -554,7 +548,20 @@ poll_poll(pollObject *self, PyObject *args)
             return NULL;
         }
 
-        deadline = _PyTime_GetMonotonicClock() + timeout;
+        if (timeout >= 0) {
+            deadline = _PyTime_GetMonotonicClock() + timeout;
+        }
+    }
+
+    /* On some OSes, typically BSD-based ones, the timeout parameter of the
+       poll() syscall, when negative, must be exactly INFTIM, where defined,
+       or -1. See issue 31334. */
+    if (ms < 0) {
+#ifdef INFTIM
+        ms = INFTIM;
+#else
+        ms = -1;
+#endif
```

Comment:
This normalization must be moved at the beginning of the loop below, since ms is recomputed if poll() is interrupted by a signal (EINTR).

Reply:
@haypo where do you mean exactly? `ms` is recalculated only if `timeout` is non-negative and `ms` cannot be negative with a non-negative `timeout`.

Reply:
I agree with @volans-. No changes are needed here.

Reply:
Right.